packages/lib-webpack/src/webpack/DynamicRemotePlugin.ts

```diff
@@ -0,0 +1,167 @@
+   *
+   * Default value: empty object.
+   */
+  sharedModules: WebpackSharedModules;
+}>;
+
+export class DynamicRemotePlugin implements webpack.WebpackPluginInstance {
+  private readonly pluginMetadata: PluginBuildMetadata;
+
+  private readonly extensions: Extension[];
+
+  private readonly sharedModules: WebpackSharedModules;
+
+  constructor(options: DynamicRemotePluginOptions = {}) {
+    const adaptedOptions: Required<DynamicRemotePluginOptions> = {
+      pluginMetadata: options.pluginMetadata ?? 'plugin.json',
+      extensions: options.extensions ?? 'extensions.json',
+      sharedModules: options.sharedModules ?? {},
+    };
+
+    // Load and validate plugin metadata
+    if (typeof adaptedOptions.pluginMetadata === 'string') {
+      this.pluginMetadata = loadPluginMetadata(adaptedOptions.pluginMetadata);
+    } else {
+      this.pluginMetadata = validatePluginMetadata(adaptedOptions.pluginMetadata);
+    }
+
+    // Load and validate extensions contributed by the plugin
+    if (typeof adaptedOptions.extensions === 'string') {
+      this.extensions = loadExtensions(adaptedOptions.extensions);
+    } else {
+      this.extensions = validateExtensions(adaptedOptions.extensions);
+    }
+
+    // Create plugin vs. host application shared module configuration
+    this.sharedModules = {
+      ...adaptedOptions.sharedModules,
+      ...getSharedModulesWithStrictSingletonConfig(['react', 'redux']),
+    };
+  }
+
+  apply(compiler: webpack.Compiler) {
+    const logger = compiler.getInfrastructureLogger(DynamicRemotePlugin.name);
+    const containerName = this.pluginMetadata.name;
+
+    if (!compiler.options.output.publicPath) {
+      logger.warn(
+        'output.publicPath is not defined, which may cause plugin assets to not load properly in the browser',
+      );
+    }
+
+    if (!compiler.options.output.uniqueName) {
+      logger.info(`output.uniqueName will be set to ${containerName}`);
+      compiler.options.output.uniqueName = containerName;
+    }
+
+    // Generate webpack federated module container assets
+    new webpack.container.ModuleFederationPlugin({
+      name: containerName,
+      library: {
+        type: 'jsonp',
+        name: remoteEntryCallback,
+      },
+      filename: remoteEntryScript,
+      exposes: _.mapValues(
+        this.pluginMetadata.exposedModules || {},
+        (moduleRequest, moduleName) => ({
+          import: moduleRequest,
+          name: `exposed-${moduleName}`,
+        }),
+      ),
+      shared: this.sharedModules,
+    }).apply(compiler);
+
+    // Generate plugin manifest
+    new GenerateManifestPlugin({
+      name: this.pluginMetadata.name,
+      version: this.pluginMetadata.version,
+      extensions: this.extensions,
+    }).apply(compiler);
+
+    // Post-process container entry generated by ModuleFederationPlugin
+    new PatchContainerEntryPlugin(this.pluginMetadata.name).apply(compiler);
```

Comment:
I would add option to bypass this patch, the extension could eventually write their own webpack plugin and sew them together in different way.

Reply:
Future feature ticket: https://issues.redhat.com/browse/HAC-767